### PR TITLE
Add a ruby script to update the Xcode project to avoid requiring manu…

### PR DESCRIPTION
**Pull request by jonspock**
…al edit.
Dev Team Id needs correction

### CODE REVIEW

File: ios/update_project.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+# This should be the Xcode ID for the team
+dev_team = "MZ11Z1Z11Z";
+
+project_path = "iOS/Electron-Cash.xcodeproj";
+
+# Create project object
+project = Xcodeproj::Project.open(project_path);
+
+lib = '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/lib/libxml2.tbd'
```
Comment (cculianu) [inline ios/update_project.rb]: I wonder if you can use `xcode-select -p` here to derive the base of the path...?

**Comment by cculianu** [comment]
Interesting. I know not any ruby. Didn't realize it had such a good lib for modifying xcode projects.  The Python resources for this are a bit lackluster.

I added a comment about perhaps auto-detecting the Xcode base path (not that it really can differ much).

Also -- would it be too much to ask/prompt user to paste in the dev team?  Maybe even auto-pick-it-up from current project setting?

Overall I'm glad this auto-adds the libxml2.tbd.

Note that even with this user still would need to go in and modify project settings to set all the "disable strip" and "export symbols" and all the other stuff the app needs...

**Comment by jonspock** [comment]
Good points. I thought those might actually come up. Initially I just wanted to open the door to automating this a little more. Will try to add those capabilities.  Lib path also doesn't work for the simulator but can't remember if the code even works on the simulator. I think i may have had issues with python lib stuff.
Auto picking up dev team from current project setting? not sure exactly what you mean. I was sort of thinking it could be held in a local .settings or other dot file and used if the file existed. Otherwise, prompt and then save to said dot file for next time.  That file should be ignored by git. 
Perhaps the other needed settings could be added later. When I built myself I guess I skipped the `disable strip` and `export symbols` stuff.

**Comment by cculianu** [comment]
> Good points. I thought those might actually come up. Initially I just wanted to open the door to automating this a little more. 

I'm grateful for that. Feel free to flesh this out more and even call this script from the main script `make_ios_project.sh`!  I just have to make sure it "does no damage" first (ie that the simulator still works and that the libxml2.tbd it linked doesn't break it somehow). 

> Lib path also doesn't work for the simulator but can't remember if the code even works on the simulator.

This code does indeed run great in the simulator.  I wouldn't know offhand how the libpath differs there -- doesn't xcode automagically figure out the corresponding simulator lib? Perhaps not.. Hmm.

> Auto picking up dev team from current project setting? not sure exactly what you mean.  

I dunno maybe the dev went in as the first thing they did and set the signer.  The dev team thing isn't that important.  I may actually ask you to take that out -- since you basically should set that in Xcode anyway -- most devs do "Xcode managed signing" and they can enable that.

Or are you envisioning some utopian future where this entire process is scriptable and requires no devs to fire up Xcode?

> Otherwise, prompt and then save to said dot file for next time. That file should be ignored by git.

I guess that works in a pinch.

> When I built myself I guess I skipped the disable strip and export symbols stuff.

Export symbols is absolutely necessary now in the latest incarnation; basically I added libsecp256k1 which is compiled INTO the main executable -- and the Python scripts `dlopen` the main executable to find the symbols for it via ctypes!  The reason it's done this way is because the basic electron cash codebase expects libsecp256k1 to be a lib so I just wanted to re-use that existing code (see `lib/seckp256k1.py`).

Last year; I found that not doing disable strip led to it not working right or being reject by app store (I forget which).  But YMMV.  It's a bit fuzzy now why I had to do that.  I'm pretty sure all my recommendations are still relevant and come as a result of personal pain with the App Store process and are not capricious in the least.

**Comment by jonspock** [comment]
OK. Added a couple of more commits to handle getting the Xcode in use and setting various attributes in Xcode. Removed use of dev team

**Comment by cculianu** [comment]
Ok, I finally got around to testing this.  I got this error with a fresh clone, fairly fresh install on python3 3.6.9, etc on macOS Mojave 10.14.6, with all Python requirements present.

```
/Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project/object/helpers/build_settings_array_settings_by_object_version.rb:27:in `<module:BuildSettingsArraySettingsByObjectVersion>': undefined method `to_set' for #<Array:0x00007faee924acf8> (NoMethodError)
Did you mean?  to_s
	from /Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project/object/helpers/build_settings_array_settings_by_object_version.rb:8:in `<class:XCBuildConfiguration>'
	from /Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project/object/helpers/build_settings_array_settings_by_object_version.rb:4:in `<module:Object>'
	from /Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project/object/helpers/build_settings_array_settings_by_object_version.rb:3:in `<class:Project>'
	from /Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project/object/helpers/build_settings_array_settings_by_object_version.rb:2:in `<module:Xcodeproj>'
	from /Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project/object/helpers/build_settings_array_settings_by_object_version.rb:1:in `<top (required)>'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project/object/build_configuration.rb:194:in `<top (required)>'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project/object.rb:521:in `<top (required)>'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /Library/Ruby/Gems/2.3.0/gems/xcodeproj-1.5.8/lib/xcodeproj/project.rb:5:in `<top (required)>'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from update_project.rb:8:in `<main>'
```

My ruby-foo is non-existant.  I know not anything about this language or its package system.  Is this some missing requirement?  Or some other bug?

Note that I don't use Brew (which maybe updates ruby?). I use MacPorts.  I also feel MacPorts is superior so we need to support both brew and MacPorts (since I developed this originally, and I maintain it -- I get to call the shots there, ha!).

Let me know if there is anything I would need to install to get the ruby prereqs up to speed on a non-brew system... (and we can add that to the early part of the `make_ios_project.sh` setup)?

**Comment by jonspock** [comment]
OK. I'm not a ruby programmer either by any means.  I have homebrew ruby 2.6.3 on one system and the /usr/bin/ruby 2.3.7p456 (2018-03-28 revision 63024) [universal.x86_64-darwin18] (natively installed - Mojave) on another. Neither seems to have a problem with the script when run standalone. Sorry this is an issue.   I do see xcodeproj under /Library/Ruby/Gems/2.3.0/gems also but it is version  xcodeproj-1.12.0 instead of your 1.5.8

**Comment by cculianu** [comment]
I wouldn't be surprised if HomeBrew installed that different package. I recently upgraded this system to Mojave and reinstalled Xcode and everything is up to date.  Yeah that must be a HomeBrew thing.

Well we can't call this script unconditionally then. Either:

1. If you  want -- make the calling script (`make_ios_project.sh` ) figure out if HomeBrew or if MacPorts (probably check for 'brew' command and/or 'ports' command?!), or;

2. Make the calling script not actually run this script but just advise the user that if on HomeBrew, they can run this script and it will do all that stuff for them.

Not sure which of (1) or (2) I prefer.  I hate error messages or software that doesn't work right.. so if there's a risk it will stack trace on people -- would prefer (2) honestly to keep the perception of quality high.

What do you think?

**Comment by jonspock** [comment]
Homebrew puts binaries in /usr/local and AFAIK doesn't change things in /usr/bin. 
Have you tried 
> gem update --system
> gem update xcodeproj
If ruby is still an issue, I'd probably just want to make the script optional as figuring out brew vs macports would be a puzzle. Let me know please

**Comment by cculianu** [comment]
MacOS GateKeeper doesn't permit updates. Note that I ran with `sudo`:

```
calin@CalinMBP2016 ~/src/ec/ec-test/ios $ sudo gem update --system
Updating rubygems-update
Fetching: rubygems-update-3.0.6.gem (100%)
Successfully installed rubygems-update-3.0.6
Parsing documentation for rubygems-update-3.0.6
Installing ri documentation for rubygems-update-3.0.6
Installing darkfish documentation for rubygems-update-3.0.6
Done installing documentation for rubygems-update after 47 seconds
Parsing documentation for rubygems-update-3.0.6
Done installing documentation for rubygems-update after 0 seconds
Installing RubyGems 3.0.6
ERROR:  While executing gem ... (Errno::EPERM)
    Operation not permitted @ rb_sysopen - /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/gem
calin@CalinMBP2016 ~/src/ec/ec-test/ios $ sudo gem update xcodeproj
Updating installed gems
Updating xcodeproj
Fetching atomos-0.1.3.gem
Fetching xcodeproj-1.12.0.gem
Fetching nanaimo-0.2.6.gem
Successfully installed atomos-0.1.3
Successfully installed nanaimo-0.2.6
ERROR:  While executing gem ... (Gem::FilePermissionError)
    You don't have write permissions for the /usr/bin directory.
```

**Comment by cculianu** [comment]
It still crashes, same stack trace.

Definitely make the script optional then.  I'm super curious how you have a different setup than me. My system is pretty cleanly set up.  I didn't do anything funny other than:

1. install Mojave, latest updates
2. Install latest Xcode (non-beta; release version)
3. Install Xcode command line tools
4. Install MacPorts and requisite things (all stuff it installs is in `/opt` and it doesn't/can't touch /usr/bin).

Very curious indeed.

I'd say for now make the script optional unless you have some other/more ideas?

**Comment by jonspock** [comment]
I appear to have the same system as you except for MacPorts/Homebrew differences but still can't explain the behavioral difference.  So update_project.rb can be run optionally now

File: ios/make_ios_project.sh [hunk 2]
```diff
@@ -251,14 +251,23 @@ fi
...
 echo '**************************************************************************'
 echo '*                                                                        *'
 echo '*   Operation Complete. An Xcode project has been generated in "iOS/"    *'
 echo '*                                                                        *'
 echo '**************************************************************************'
+
+echo '  IMPORTANT!'
+echo '        Now you either need to manually add the library libxml2.tbd to the '
+echo '        project under "General -> Linked Frameworks and Libraries" or successfully '
+echo '        run the update_project.rb script to do it for you '
```
Comment (cculianu) [inline ios/make_ios_project.sh]: I know this seems ridiculous but if you don't do it I will have to do it in a subsequent commit. Would you mind ending this sentence with a period `.`?  If you don't do it I will after merge.. thus adding 1 extra commit so.. hoping you don't mind doing it here to this PR. :)

Comment (cculianu) [reply ios/make_ios_project.sh]: Eg: `run the update_project.rb script to do it for you.`  <-- period. :)

**Comment by cculianu** [comment]
This looks great. I added a tiny tiny stupid nit comment. Ha!

Will merge after that.. thanks man! Great stuff -- I hope this script helps Brew people.

Yes, I also don't understand why we are experiencing difference.  I wonder if there are alternate package paths for ruby that Brew uses or adds -- thus Brew added some installed packages in `/usr/local` perhaps?

Anyway let me know if you can do that nit and I'll merge.  If I don't hear from you for a few days I'll merge anyway and just do the nit myself.

Thanks again.

File: ios/make_ios_project.sh (hunk 2 above)
Comment (jonspock) [reply ios/make_ios_project.sh]: OK. Added. No problem.
Also don't understand the differences.
Glad to help and see this merged.
I'm presuming you'll squash, but I can rebase/squash if needed

**Comment by cculianu** [comment]
Yeah, no worries. I'll just squash. :) Thanks for the nit edit. :)

**Pull request merged**